src/app/qgisapp.cpp

```diff
@@ -16046,6 +16046,61 @@ void QgisApp::triggerCrashHandler()
 #endif
 }
 
+void QgisApp::addTabifiedDockWidget( Qt::DockWidgetArea area, QDockWidget *dockWidget, const QString &onTopOf )
+{
+  QList< QDockWidget * > dockWidgetsInArea;
+  const auto dockWidgets = findChildren< QDockWidget * >();
+  for ( QDockWidget *w : dockWidgets )
+  {
+    if ( w->isVisible() && dockWidgetArea( w ) == area )
+    {
+      dockWidgetsInArea << w;
+    }
+  }
+
+  addDockWidget( area, dockWidget );  // First add the dock widget, then attempt to tabify
+  if ( dockWidgetsInArea.length() > 0 )
+  {
+    QDockWidget *onTopOfDockWidget = nullptr;
+    if ( !onTopOf.isEmpty() )
+    {
+      for ( QDockWidget *cw : dockWidgetsInArea )
+      {
+        if ( cw->windowTitle() == onTopOf )
+        {
+          onTopOfDockWidget = cw;
+          break;
+        }
+      }
+    }
+
+    if ( !onTopOfDockWidget )
+    {
+      onTopOfDockWidget = dockWidgetsInArea.at( 0 );  // Last resort
+    }
+    tabifyDockWidget( onTopOfDockWidget, dockWidget );
+    qApp->processEvents();  // Give the app a while to refresh controls
```

Comment:
This is dangerous to call. I'd suggest making the auto focus behaviour a manual step and don't perform it automatically here.

Reply:
Note that if you use QgsDockWidget it's easy to do via setUserVisible

Reply:
Wow! It was easier than I was suggesting, and also effective. Thanks!